### PR TITLE
Changes to stackwidget to allow for new ComponentTreeView

### PR DESCRIPTION
**Pull request by keenfoong**
Changes to layouts, and stackwidget for new component tree widget. Also to separate stack widget items from layouts

### CODE REVIEW

**Comment by keenfoong** [comment]
My bad, thought we needed the code out quickly for you to test if you wanted. I'll fix this up today.

File: zoo/libs/pyqt/widgets/buttonlist.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# Embedded file name: D:\git\zootools_pro\zoocore\zoo\libs\pyqt\widgets\buttonlist.py
+from qt import QtWidgets, QtCore
+
+class ButtonList(QtWidgets.QWidget):
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/buttonlist.py]: Would like to see a screenshot of this widget

File: zoo/libs/pyqt/widgets/buttonlist.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    """A nice vertical button list with an optional header and with a searchable feature. """
+
+    def __init__(self, parent = None, title = '', showTitle = True, showSearch = True):
+        super(ButtonList, self).__init__(parent=parent)
+        self.mainLayout = QtWidgets.QVBoxLayout()
+        self.mainLayout.setContentsMargins(2, 4, 2, 4)
+        self.titleLbl = QtWidgets.QLabel(title)
+        self.searchEdit = QtWidgets.QLineEdit()
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/buttonlist.py]: you should be setting the parent widget

File: zoo/libs/pyqt/widgets/buttonlist.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+# Embedded file name: D:\git\zootools_pro\zoocore\zoo\libs\pyqt\widgets\buttonlist.py
+from qt import QtWidgets, QtCore
+
+class ButtonList(QtWidgets.QWidget):
+    """A nice vertical button list with an optional header and with a searchable feature. """
+
+    def __init__(self, parent = None, title = '', showTitle = True, showSearch = True):
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/buttonlist.py]: Dont use strings as the default argument value , it's a mutable type

File: zoo/libs/pyqt/widgets/buttonlist.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    def initUi(self):
+        self.searchEdit.setPlaceholderText('Search...')
+        self.toolBar.addWidget(self.searchEdit)
+        self.mainLayout.addWidget(self.titleLbl)
+        self.mainLayout.addLayout(self.toolBar)
+        self.setLayout(self.mainLayout)
+        self.mainLayout.addWidget(self.buttonList)
+
+    def connections(self):
+        """ Test connection """
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/buttonlist.py]: test connection?

File: zoo/libs/pyqt/widgets/stackwidget.py [hunk 5]
```diff
@@ -79,41 +84,86 @@ def initUi(self):
         self.setLayout(mainLayout)
 
     def connections(self):
+        """
+        Connections for the buttons and the search edit
+        :return:
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/stackwidget.py]: I honestly wouldn't bother adding :return: when the func returns None. If the docstring doesn't specify a return it's expected to be None.

File: zoo/libs/pyqt/widgets/stackwidget.py [hunk 6]
```diff
@@ -79,41 +84,86 @@ def initUi(self):
...
+        """
+        Clear all the items in the stack
+        :return:
+        """
         self.stackTableWgt.clearStack()
 
     def addStackItem(self, item):
+        """
+        Add item to the StackTableWidget
+        :param item: StackItem to add to the table
```
Comment (dsparrow27) [inline zoo/libs/pyqt/widgets/stackwidget.py]: Make sure you have a space between the header description and the body, this is so our documentation tool will render it correctly
eg.

    """
    Some one line description

    Longer description

    :param item
    """

**Pull request merged**